File: code/python/tree/avl.py
from no_element import NoElement


class AVL:
    """
    AVL trees where the heights of the two child subtrees of any node
    differ by at most one.
    """

    class Node:
        def __init__(self, key, height=0, left=None, right=None):
            self.key = key
            self.height = height
            self.left = left
            self.right = right

    def __init__(self):
        self._root = None

    def is_empty(self):
        return self._root is None

    @staticmethod
    def _get_height(node: Node):
        if node is None:
            return -1
        else:
            return node.height

    def height(self):
        return AVL._get_height(self._root)

    @staticmethod
    def _get_balance_factor(node: Node):
        if node is None:
            return 0
        else:
            return AVL._get_height(node.left) - AVL._get_height(node.right)

    def get(self, key):
        def _get(x: AVL.Node):
            if x is None or key == x.key:
                return x
            if key < x.key:
                return _get(x.left)
            else:
                return _get(x.right)
        if key is None:
            raise KeyError
        return _get(self._root)

    def contains(self, key):
        return self.get(key) is not None

    @staticmethod
    def _right_rotate(y: Node):
        x = y.left
        y.left = x.right
        x.right = y
        y.height = max(AVL._get_height(y.left), AVL._get_height(y.right)) + 1
        x.height = max(AVL._get_height(x.left), AVL._get_height(x.right)) + 1
        return x

    @staticmethod
    def _left_rotate(x: Node):
        y = x.right
        x.right = y.left
        y.left = x
        x.height = max(AVL._get_height(x.left), AVL._get_height(x.right)) + 1
        y.height = max(AVL._get_height(y.left), AVL._get_height(y.right)) + 1
        return y

    def put(self, key):
        def _put(x: AVL.Node):
            if x is None:
                return AVL.Node(key)
            elif key < x.key:
                x.left = _put(x.left)
            elif key > x.key:
                x.right = _put(x.right)
            # update the height
            x.height = max(AVL._get_height(x.left), AVL._get_height(x.right)) + 1
            # get the balance factor
            bf = AVL._get_balance_factor(x)
            if bf > 1 and key < x.left.key:  # case 1
                return AVL._right_rotate(x)
            if bf < -1 and key > x.right.key:  # case 2
                return AVL._left_rotate(x)
            if bf > 1 and key > x.left.key:  # case 3
                x.left = AVL._left_rotate(x.left)
                return AVL._right_rotate(x)
            if bf < -1 and key < x.right.key:  # case 4
                x.right = AVL._right_rotate(x.right)
                return AVL._left_rotate(x)
            return x

        self._root = _put(self._root)
        assert self.is_avl()

    @staticmethod
    def _min(x: Node):
        # assume x is not None
        if x.left is None:
            return x
        return AVL._min(x.left)

    def min(self):
        if self.is_empty():
            raise NoElement
        return AVL._min(self._root).key

    @staticmethod
    def _remove_min(x: Node):
        # assume x is not None
        if x.left is None:
            return x.right
        x.left = AVL._remove_min(x.left)
        x.height = max(AVL._get_height(x.left), AVL._get_height(x.right)) + 1
        return x

    def remove(self, key):
        def _remove(x: AVL.Node):
            if x is None:
                return None
            if key < x.key:
                x.left = _remove(x.left)
            elif key > x.key:
                x.right = _remove(x.right)
            else:
                if x.right is None:
                    return x.left
                if x.left is None:
                    return x.right
                t = x
                x = AVL._min(t.right)
                x.right = AVL._remove_min(t.right)
                x.left = t.left
            x.height = max(AVL._get_height(x.left), AVL._get_height(x.right)) + 1
            bf = AVL._get_balance_factor(x)
            if bf > 1 and AVL._get_balance_factor(x.left) >= 0:  # case 1: left-left
                return AVL._right_rotate(x)
            if bf < -1 and AVL._get_balance_factor(x.right) <= 0:  # case 2: right-right
                return AVL._left_rotate(x)
            if bf > 1 and AVL._get_balance_factor(x.left) < 0:  # case 3: left-right
                x.left = AVL._left_rotate(x.left)
                return AVL._right_rotate(x)
            if bf < -1 and AVL._get_balance_factor(x.right) > 0:  # case 4: right-left
                x.right = AVL._right_rotate(x.right)
                return AVL._left_rotate(x)
            return x
        self._root = _remove(self._root)
        assert self.is_avl()

    @staticmethod
    def _is_avl(x: Node):
        if x is None:
            return True
        return abs(AVL._get_balance_factor(x)) < 2 and AVL._is_avl(x.left) and AVL._is_avl(x.right)

    def is_avl(self):
        return AVL._is_avl(self._root)

File: code/java/tree/src/main/java/org/swufe/AVL.java
package org.swufe;

import java.util.NoSuchElementException;

public class AVL <Key extends Comparable<Key>> {
    private static class Node<Key extends Comparable<Key>> {
        Key key;
        int height;
        Node<Key> left;
        Node<Key> right;

        Node(Key key) {
            this.key = key;
            height = 0;
        }
    }

    private Node<Key> root;

    public AVL() {
        root = null;
    }

    public boolean isEmpty() {
        return root == null;
    }

    private Node<Key> get(Node<Key> x, Key key) {
        if (x == null || key.compareTo(x.key) == 0) {
            return x;
        }
        if (key.compareTo(x.key) < 0) {
            return get(x.left, key);
        } else  {
            return get(x.right, key);
        }
    }

    public Node<Key> get(Key key) {
        if (key == null) {
            throw new IllegalArgumentException();
        }
        return get(root, key);
    }

    public boolean contains(Key key) {
        return get(key) != null;
    }

    private int getHeight(Node<Key> x) {
        if (x == null) return 0;
        return x.height;
    }

    public int getHeight() {
        return getHeight(root);
    }

    private int getBalanceFactor(Node<Key> x) {
        if (x == null) return 0;
        return getHeight(x.left) - getHeight(x.right);
    }

    private Node<Key> rotateRight(Node<Key> x) {
        assert x != null && x.left != null;
        Node<Key> y = x.left;
        x.left = y.right;
        y.right = x;
        x.height = Integer.max(getHeight(x.left), getHeight(x.right)) + 1;
        y.height = Integer.max(getHeight(y.left), getHeight(y.right)) + 1;
        return y;
    }

    private Node<Key> rotateLeft(Node<Key> y) {
        assert y != null && y.right != null;
        Node<Key> x = y.right;
        y.right = x.left;
        x.left = y;
        y.height = Integer.max(getHeight(y.left), getHeight(y.right)) + 1;
        x.height = Integer.max(getHeight(x.left), getHeight(x.right)) + 1;
        return x;
    }

    private Node<Key> put(Node<Key> x, Key key) {
        if (x == null) {
            return new Node<>(key);
        }
        if (key.compareTo(x.key) < 0) {
            x.left = put(x.left, key);
        } else if (key.compareTo(x.key) > 0) {
            x.right = put(x.right, key);
        }
        x.height = Integer.max(getHeight(x.left), getHeight(x.right)) + 1;
        int bf = getBalanceFactor(x);
        if (bf > 1 && key.compareTo(x.left.key) < 0) {
            return rotateRight(x);
        }
        if (bf < -1 && key.compareTo(x.right.key) > 0) {
            return rotateLeft(x);
        }
        if (bf > 1 && key.compareTo(x.left.key) > 0) {
            x.left = rotateLeft(x.left);
            return rotateRight(x);
        }
        if (bf < -1 && key.compareTo(x.right.key) < 0) {
            x.right = rotateRight(x.right);
            return rotateLeft(x);
        }
        return x;
    }

    public void put(Key key) {
        root = put(root, key);
        assert isAVL();
    }

    public Node<Key> min(Node<Key> x) {
        assert x != null;
        if (x.left == null) return x;
        return min(x.left);
    }

    public Key min() {
        if (isEmpty()) throw new NoSuchElementException();
        return min(root).key;
    }

    private Node<Key> removeMin(Node<Key> x) {
        assert x != null;
        if (x.left == null) return x.right;
        x.left = removeMin(x.left);
        x.height = Integer.max(getHeight(x.left), getHeight(x.right)) + 1;
        return x;
    }

    private Node<Key> remove(Node<Key> x, Key key) {
        if (x == null) return null;
        if (key.compareTo(x.key) < 0) {
            x.left = remove(x.left, key);
        } else if (key.compareTo(x.key) > 0) {
            x.right = remove(x.right, key);
        } else {
            if (x.right == null) return x.left;
            if (x.left == null) return x.right;
            Node<Key> t = x;
            x = min(t.right);
            x.right = removeMin(t.right);
            x.left = t.left;
        }
        x.height = Integer.max(getHeight(x.left), getHeight(x.right)) + 1;
        int bf = getBalanceFactor(x);
        if (bf > 1 && getBalanceFactor(x.left) >= 0) {
            return rotateRight(x);
        }
        if (bf < -1 && getBalanceFactor(x.right) <= 0) {
            return rotateLeft(x);
        }
        if (bf > 1 && getBalanceFactor(x.left) < 0) {
            x.left = rotateLeft(x.left);
            return rotateRight(x);
        }
        if (bf < -1 && getBalanceFactor(x.right) > 0) {
            x.right = rotateRight(x.right);
            return rotateLeft(x);
        }
        return x;
    }

    public void remove(Key key) {
        root = remove(root, key);
        assert isAVL();
    }

    private boolean isAVL(Node<Key> x) {
        if (x == null) return true;
        return Math.abs(getBalanceFactor(x)) < 2 && isAVL(x.left) && isAVL(x.right);
    }

    private boolean isAVL() {
        return isAVL(root);
    }
}
